src/core.jl

```diff
@@ -19,3 +19,187 @@ immutable ImageEdge
     index2::Int
     weight::Float64
 end
+
+"""
+    G, vert_map = region_adjacency_graph(seg, weight_fn)
+
+Constructs a region adjacency graph from the `SegmentedImage`. It returns the RAG
+along with a label->vertex map. `weight_fn` is used to assign weights to edges.
+
+    weight_fn(label1, label2)
+
+Returns a Float64 corresponding to the weight of the edge between label1 and label2.
+
+"""
+function region_adjacency_graph(s::SegmentedImage, weight_fn::Function)
+
+    function neighbor_regions!(n::Set{Int}, visited::AbstractArray, s::SegmentedImage, I::CartesianIndex)
+        R = CartesianRange(indices(s.image_indexmap))
+        I1 = one(CartesianIndex{ndims(visited)})
+        Ibegin, Iend = first(R), last(R)
+        t = Vector{CartesianIndex{ndims(visited)}}()
+        push!(t, I)
+
+        while !isempty(t)
+            temp = pop!(t)
+            visited[temp] = true
+            for J in CartesianRange(max(Ibegin, temp-I1), min(Iend, temp+I1))
+                if s.image_indexmap[temp] != s.image_indexmap[J]
+                    push!(n,s.image_indexmap[J])
+                elseif !visited[J]
+                    push!(t,J)
+                end
+            end
+        end
+        n
+    end
+
+    visited  = similar(dims->fill(false,dims), indices(s.image_indexmap))    # Array to mark the pixels that are already visited
+    G        = SimpleWeightedGraph()                                         # The region_adjacency_graph
+    vert_map = Dict{Int,Int}()                                               # Map that stores (label, vertex) pairs
+
+    # add vertices to graph
+    add_vertices!(G,length(s.segment_labels))
+
+    # setup `vert_map`
+    for (i,l) in enumerate(s.segment_labels)
+        vert_map[l] = i
+    end
+
+    # add edges to graph
+    for p in CartesianRange(indices(s.image_indexmap))
+        if !visited[p]
+            n = Set{Int}()
+            neighbor_regions!(n, visited, s, p)
+            for i in n
+                add_edge!(G, vert_map[s.image_indexmap[p]], vert_map[i], weight_fn(s.image_indexmap[p], i))
+            end
+        end
+    end
+    G, vert_map
+end
+
+
+"""
+    new_seg = rem_segment(seg, label, diff_fn)
+
+Removes the segment having label `label` and returns the new `SegmentedImage`.
+For more info, see [`remove_segment!`](@ref)
+
+"""
+rem_segment(s::SegmentedImage, args...) = rem_segment!(deepcopy(s), args...)
+
+"""
+    rem_segment!(seg, label, diff_fn)
+
+In place removal of the segment having label `label`, replacing it with the neighboring
+segment having least `diff_fn` value.
+
+    diff_fn(rem_label, neigh_label)
+
+A difference measure between label to be removed and its neighbors. `isless` must be
+defined for this measure.
```

Comment:
This line isn't very clear. Maybe show the output, `d = diff_fn(rem_label, neigh_label)` and then say that `isless` must be defined for objects of the type of `d`.

Also an issue below.